src/doc/selected_frames.h

```diff
@@ -16,126 +16,136 @@
+      return !operator==(it);
+    }
 
-      const_reverse_iterator& operator++() {
-        if (m_frame < 0)
-          m_frame = m_it->toFrame;
+  private:
+    mutable frames::Ranges::const_iterator m_it;
+    mutable frame_t m_frame;
+  };
 
-        if (m_frame > m_it->fromFrame)
-          --m_frame;
-        else {
-          m_frame = -1;
-          ++m_it;
-        }
+  class const_reverse_iterator {
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = frame_t;
+    using difference_type = std::ptrdiff_t;
+    using pointer = frame_t*;
+    using reference = frame_t&;
+
+    const_reverse_iterator(const frames::Ranges::const_reverse_iterator& it)
+      : m_it(it), m_frame(-1) {
+    }
 
-        return *this;
-      }
+    const_reverse_iterator& operator++() {
+      if (m_frame < 0)
+        m_frame = m_it->toFrame;
 
-      frame_t operator*() const {
-        if (m_frame < 0)
-          m_frame = m_it->toFrame;
-        return m_frame;
-      }
+      int step = (m_it->fromFrame > m_it->toFrame) -
+                 (m_it->fromFrame < m_it->toFrame);
 
-      bool operator==(const const_reverse_iterator& o) const {
-        return (m_it == o.m_it && m_frame == o.m_frame);
+      if ((step < 0 && m_frame > m_it->fromFrame) ||
+          (step > 0 && m_frame < m_it->fromFrame))
+        m_frame += step;
+      else {
+        m_frame = -1;
+        ++m_it;
       }
 
-      bool operator!=(const const_reverse_iterator& it) const {
-        return !operator==(it);
-      }
+      return *this;
+    }
 
-    private:
-      mutable Ranges::const_reverse_iterator m_it;
-      mutable frame_t m_frame;
-    };
+    frame_t operator*() const {
+      if (m_frame < 0)
+        m_frame = m_it->toFrame;
+      return m_frame;
+    }
 
-    class Reversed {
-    public:
-      typedef const_reverse_iterator const_iterator;
+    bool operator==(const const_reverse_iterator& o) const {
+      return (m_it == o.m_it && m_frame == o.m_frame);
+    }
 
-      const_iterator begin() const { return m_selectedFrames.rbegin(); }
-      const_iterator end() const { return m_selectedFrames.rend(); }
+    bool operator!=(const const_reverse_iterator& it) const {
+      return !operator==(it);
+    }
 
-      Reversed(const SelectedFrames& selectedFrames)
-        : m_selectedFrames(selectedFrames) {
-      }
+  private:
+    mutable frames::Ranges::const_reverse_iterator m_it;
+    mutable frame_t m_frame;
+  };
 
-    private:
-      const SelectedFrames& m_selectedFrames;
-    };
+  template<typename T>
+  class Reversed {
+  public:
+    typedef const_reverse_iterator const_iterator;
+
+    const_iterator begin() const { return m_frames.rbegin(); }
+    const_iterator end() const { return m_frames.rend(); }
+
+    Reversed(const T& frames)
+      : m_frames(frames) {
+    }
 
+  private:
+    const T& m_frames;
+  };
+
+} // namespace frames
+
+  using namespace frames;
```

Comment:
Creating a namespace and immediately using it has the same result that including everything in the parent namespace. In this way if we are in (or include) `doc` namespace we have `const_iterator` type available as a global type (which is not desirable).

One easy alternative (not the most elegant one but it works) is to change the namespace to a struct:

```c++
  struct frames {
    // ...
  }; // struct frames

  class SelectedFrames : public frames {
  public:
    // ...
```

Reply:
Oh I see. I think I will move the doc::frames namespace to an independent header file. Thank you!

Reply:
I think the problem is still persistent, the issue here is `using namespace doc::frames;`, not the header file name or where the namespace is defined.

Reply:
Why do you say it? Maybe I'm getting it wrong, or I'm testing it wrong.
To check if this works I'm doing the following:
1. Go to layer_frame_comboboxes.cpp file (actually any file that includes selected_frames.h)
2. In some method try to declare:
`doc::frames::Ranges r;` <-- compiles
`doc::Ranges r;` <-- doesn't compile

I'm using the Ranges type, but with the const_iterator class happens the same given appropriate ctor parameters.

Before moving the namespace to its own header, and more importantly (because this made it actually work), moving the `using namespace doc::frames` outside of the doc namespace, I was able to use doc::Ranges or doc::const_iterator.

Reply:
Add this line in `sprite.cpp`:
```diff
diff --git a/src/doc/sprite.cpp b/src/doc/sprite.cpp
index b26cd0e77..861ff33a2 100644
--- a/src/doc/sprite.cpp
+++ b/src/doc/sprite.cpp
@@ -468,6 +468,8 @@ RgbMap* Sprite::rgbMap(const frame_t frame,
 
 void Sprite::addFrame(frame_t newFrame)
 {
+  const_iterator it;
+
   setTotalFrames(m_frames+1);
 
   frame_t to = std::max(1, newFrame);
```
You will get this error:
```
src/doc/sprite.cpp:471:18: error: no matching constructor for initialization of 'const_iterator'
  const_iterator it;
                 ^
src/doc/../doc/frames_iterators.h:30:5: note: candidate constructor not viable: requires single argument 'it', but no arguments were provided
    const_iterator(const frames::Ranges::const_iterator& it)
    ^
src/doc/../doc/frames_iterators.h:21:9: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 0 were provided
  class const_iterator {
        ^
src/doc/../doc/frames_iterators.h:21:9: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 0 were provided
```
Which means that the `const_iterator` type is available in the whole `doc` namespace.

If you make these changes:
```diff
diff --git a/src/doc/frames_iterators.h b/src/doc/frames_iterators.h
index 0a2d03b1d..830aa5f16 100644
--- a/src/doc/frames_iterators.h
+++ b/src/doc/frames_iterators.h
@@ -15,7 +15,7 @@
 
 namespace doc {
 
-namespace frames {
+struct frames {
   typedef std::vector<FrameRange> Ranges;
 
   class const_iterator {
@@ -139,7 +139,7 @@ namespace frames {
     const T& m_frames;
   };
 
-} // namespace frames
+}; // struct frames
 } // namespace doc
 
 #endif  // DOC_FRAMES_ITERATORS_H_INCLUDED
diff --git a/src/doc/frames_sequence.h b/src/doc/frames_sequence.h
index e32bd314c..a51ea217c 100644
--- a/src/doc/frames_sequence.h
+++ b/src/doc/frames_sequence.h
@@ -10,14 +10,12 @@
 
 #include "doc/selected_frames.h"
 
-using namespace doc::frames;
-
 namespace doc {
 
   // This class is based in several code of the SelectedFrames class.
   // TODO: At some point we should remove the duplicated code between these
   // classes.
-  class FramesSequence {
+  class FramesSequence : public frames {
   public:
     FramesSequence() {}
     FramesSequence(const SelectedFrames& selectedFrames);
diff --git a/src/doc/selected_frames.h b/src/doc/selected_frames.h
index 566e5df0d..785ee555a 100644
--- a/src/doc/selected_frames.h
+++ b/src/doc/selected_frames.h
@@ -11,14 +11,12 @@
 
 #include "doc/frames_iterators.h"
 
-using namespace doc::frames;
-
 namespace doc {
 
   // The FramesSequence class is based in several code of this class.
   // TODO: At some point we should remove the duplicated code between these
   // classes.
-  class SelectedFrames {
+  class SelectedFrames : public frames {
   public:
     const_iterator begin() const { return const_iterator(m_ranges.begin()); }
     const_iterator end() const { return const_iterator(m_ranges.end()); }
```
And recompile you will get this error:
````
src/doc/sprite.cpp:471:3: error: unknown type name 'const_iterator'
  const_iterator it;
  ^
```
Which means that (e.g.) the `const_iterator` is not a type published outside `FramesSequence/SelectedFrames` bounds anymore.

Reply:
I see what you mean now, thank you for clarifying. I don't really like the semantics of adding a struct to make a couple of classes inherit from it just to make the namespaces work as we want, it just feels wrong to me. There is an alternative that I was avoiding because of the use of the "frames" namespace repeated times which I don't like either, but makes me feel better in terms of semantics:

```diff
diff --git a/src/doc/frames_sequence.h b/src/doc/frames_sequence.h
index e32bd314c..488df1048 100644
--- a/src/doc/frames_sequence.h
+++ b/src/doc/frames_sequence.h
@@ -10,8 +10,6 @@
 
 #include "doc/selected_frames.h"
 
-using namespace doc::frames;
-
 namespace doc {
 
   // This class is based in several code of the SelectedFrames class.
@@ -22,10 +20,10 @@ namespace doc {
     FramesSequence() {}
     FramesSequence(const SelectedFrames& selectedFrames);
 
-    const_iterator begin() const { return const_iterator(m_ranges.begin()); }
-    const_iterator end() const { return const_iterator(m_ranges.end()); }
-    const_reverse_iterator rbegin() const { return const_reverse_iterator(m_ranges.rbegin()); }
-    const_reverse_iterator rend() const { return const_reverse_iterator(m_ranges.rend()); }
+    frames::const_iterator begin() const { return frames::const_iterator(m_ranges.begin()); }
+    frames::const_iterator end() const { return frames::const_iterator(m_ranges.end()); }
+    frames::const_reverse_iterator rbegin() const { return frames::const_reverse_iterator(m_ranges.rbegin
()); }
+    frames::const_reverse_iterator rend() const { return frames::const_reverse_iterator(m_ranges.rend());
 }
 
     std::size_t size() const;
     std::size_t ranges() const { return m_ranges.size(); }
@@ -44,7 +42,7 @@ namespace doc {
     frame_t lowestFrame() const;
 
     void displace(frame_t frameDelta);
-    Reversed<FramesSequence> reversed() const { return Reversed(*this); }
+    frames::Reversed<FramesSequence> reversed() const { return frames::Reversed(*this); }
 
     FramesSequence makeReverse() const;
     FramesSequence makePingPong() const;
@@ -61,7 +59,7 @@ namespace doc {
     bool read(std::istream& is);
 
   private:
-    Ranges m_ranges;
+    frames::Ranges m_ranges;
   };
 
 } // namespace doc
diff --git a/src/doc/selected_frames.h b/src/doc/selected_frames.h
index 566e5df0d..3eec9e975 100644
--- a/src/doc/selected_frames.h
+++ b/src/doc/selected_frames.h
@@ -11,8 +11,6 @@
 
 #include "doc/frames_iterators.h"
 
-using namespace doc::frames;
-
 namespace doc {
 
   // The FramesSequence class is based in several code of this class.
@@ -20,10 +18,10 @@ namespace doc {
   // classes.
   class SelectedFrames {
   public:
-    const_iterator begin() const { return const_iterator(m_ranges.begin()); }
-    const_iterator end() const { return const_iterator(m_ranges.end()); }
-    const_reverse_iterator rbegin() const { return const_reverse_iterator(m_ranges.rbegin()); }
-    const_reverse_iterator rend() const { return const_reverse_iterator(m_ranges.rend()); }
+    frames::const_iterator begin() const { return frames::const_iterator(m_ranges.begin()); }
+    frames::const_iterator end() const { return frames::const_iterator(m_ranges.end()); }
+    frames::const_reverse_iterator rbegin() const { return frames::const_reverse_iterator(m_ranges.rbegin
()); }
+    frames::const_reverse_iterator rend() const { return frames::const_reverse_iterator(m_ranges.rend());
 }
 
     std::size_t size() const;
     std::size_t ranges() const { return m_ranges.size(); }
@@ -40,7 +38,7 @@ namespace doc {
     frame_t lastFrame() const { return (!m_ranges.empty() ? m_ranges.back().toFrame: -1); }
 
     void displace(frame_t frameDelta);
-    Reversed<SelectedFrames> reversed() const { return Reversed(*this); }
+    frames::Reversed<SelectedFrames> reversed() const { return frames::Reversed(*this); }
 
     SelectedFrames makeReverse() const;
     SelectedFrames makePingPong() const;
@@ -57,7 +55,7 @@ namespace doc {
     bool read(std::istream& is);
 
   private:
-    Ranges m_ranges;
+    frames::Ranges m_ranges;
   };
 
 } // namespace doc
```
Anyway, if you still think the struct changes worth it, I can introduce them and then relief my semantics pains with a bit of "Agaromba" 😄 
Just let me know what is your preference. Thanks.

Reply:
Yeah, the struct is not the most elegant solution. It's a shame that we cannot use "using namespace" inside a class.

Your solution is good enough as we don't refer the iterators types outside these classes (or we use auto or range-based fors), other possibility would be adding (my initial idea but got some problems with the Reversed type and left it there):

```c++
  class SelectedFrames {
  public:
    using iterator = frames::iterator;
    // etc.
```
For each type, but we can leave this with `frames::`. I'll test this again and merge it 👍